src/main/java/appeng/util/helpers/ItemComparisonHelper.java

```diff
@@ -0,0 +1,443 @@
+						}
+					}
+
+					return true;
+				}
+
+				case 1: // ( A instanceof NBTTagByte )
+					return ( (NBTPrimitive) left ).getByte() == ( (NBTPrimitive) right ).getByte();
+
+				case 4: // else if ( A instanceof NBTTagLong )
+					return ( (NBTPrimitive) left ).getLong() == ( (NBTPrimitive) right ).getLong();
+
+				case 8: // else if ( A instanceof NBTTagString )
+					return ( (NBTTagString) left ).getString().equals( ( (NBTTagString) right ).getString() ) || ( (NBTTagString) left ).getString().equals( ( (NBTTagString) right ).getString() );
+
+				case 6: // else if ( A instanceof NBTTagDouble )
+					return ( (NBTPrimitive) left ).getDouble() == ( (NBTPrimitive) right ).getDouble();
+
+				case 5: // else if ( A instanceof NBTTagFloat )
+					return ( (NBTPrimitive) left ).getFloat() == ( (NBTPrimitive) right ).getFloat();
+
+				case 3: // else if ( A instanceof NBTTagInt )
+					return ( (NBTPrimitive) left ).getInt() == ( (NBTPrimitive) right ).getInt();
+
+				default:
+					return left.equals( right );
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Unordered hash of NBT Data, used to work thought huge piles fast, but ignores the order just in case MC
+	 * decided to change it... WHICH IS BAD...
+	 */
+	public int createUnorderedNbtHash( final NBTBase nbt )
+	{
+		// same type?
+		int hash = 0;
+		final byte id = nbt.getId();
+		hash += id;
+		switch( id )
+		{
+			case 10:
+			{
+				final NBTTagCompound ctA = (NBTTagCompound) nbt;
+
+				final Set<String> cA = ctA.getKeySet();
+
+				for( final String name : cA )
+				{
+					hash += name.hashCode() ^ createUnorderedNbtHash( ctA.getTag( name ) );
+				}
+
+				return hash;
+			}
+
+			case 9: // ) // A instanceof NBTTagList )
+			{
+				final NBTTagList lA = (NBTTagList) nbt;
+				hash += 9 * lA.tagCount();
+
+				final List<NBTBase> l = tagList( lA );
+				for( int x = 0; x < l.size(); x++ )
+				{
+					hash += ( (Integer) x ).hashCode() ^ createUnorderedNbtHash( l.get( x ) );
+				}
+
+				return hash;
+			}
+
+			case 1: // ( A instanceof NBTTagByte )
+				return hash + ( (NBTPrimitive) nbt ).getByte();
+
+			case 4: // else if ( A instanceof NBTTagLong )
+				return hash + (int) ( (NBTPrimitive) nbt ).getLong();
+
+			case 8: // else if ( A instanceof NBTTagString )
+				return hash + ( (NBTTagString) nbt ).getString().hashCode();
+
+			case 6: // else if ( A instanceof NBTTagDouble )
+				return hash + (int) ( (NBTPrimitive) nbt ).getDouble();
+
+			case 5: // else if ( A instanceof NBTTagFloat )
+				return hash + (int) ( (NBTPrimitive) nbt ).getFloat();
+
+			case 3: // else if ( A instanceof NBTTagInt )
+				return hash + ( (NBTPrimitive) nbt ).getInt();
+
+			default:
+				return hash;
+		}
+	}
+
+	/**
+	 * Lots of silliness to try and account for weird tag related junk, basically requires that two tags have at least
+	 * something in their tags before it wastes its time comparing them.
+	 */
+	private boolean hasSameNbtTag( final ItemStack a, final ItemStack b )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "hasSameNbtTag" is 32 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)